tree/dataframe/src/RLoopManager.cxx

```diff
@@ -359,6 +359,68 @@ RLoopManager::RLoopManager(std::unique_ptr<RDataSource> ds, const ColumnNames_t
    fDataSource->SetNSlots(fNSlots);
 }
 
+RLoopManager::RLoopManager(const ROOT::RDF::RDatasetSpec &&spec)
+   : fStartEntry(spec.fStartEntry), fEndEntry(spec.fEndEntry), fNSlots(RDFInternal::GetNSlots()),
+     fLoopType(ROOT::IsImplicitMTEnabled() ? ELoopType::kROOTFilesMT : ELoopType::kROOTFiles),
+     fNewSampleNotifier(fNSlots), fSampleInfos(fNSlots)
+{
+   if (spec.fTreeNames.size() == 1) {
+      // A TChain has a global name, that is the name of single tree
+      auto chain = std::make_shared<TChain>(spec.fTreeNames[0].c_str());
+      // The global name of the chain is also the name of each tree in the list
+      // of files that make the chain.
+      for (const auto &f : spec.fFileNameGlobs)
+         chain->Add(f.c_str());
+      SetTree(chain);
+   } else {
+      // Some other times, each different file has its own tree name, we need to
+      // reconstruct the full path to the tree in each file and pass that to
+      // TChain::Add
+      auto chain = std::make_shared<TChain>();
+      for (auto i = 0u; i < spec.fFileNameGlobs.size(); i++) {
+         const auto fullpath = spec.fFileNameGlobs[i] + "?#" + spec.fTreeNames[i];
+         chain->Add(fullpath.c_str());
+      }
+      SetTree(chain);
+   }
+
+   // Create the friends from the list of friends
+   const auto &friendNames = spec.fFriendInfo.fFriendNames;
+   const auto &friendFileNames = spec.fFriendInfo.fFriendFileNames;
+   const auto &friendChainSubNames = spec.fFriendInfo.fFriendChainSubNames;
+   const auto nFriends = friendNames.size();
+
+   for (auto i = 0u; i < nFriends; ++i) {
+      const auto &thisFriendNameAlias = friendNames[i];
+      const auto &thisFriendName = thisFriendNameAlias.first;
+      const auto &thisFriendAlias = thisFriendNameAlias.second;
+      const auto &thisFriendFiles = friendFileNames[i];
+      const auto &thisFriendChainSubNames = friendChainSubNames[i];
+
+      // Build a friend chain
+      auto frChain = std::make_unique<TChain>(thisFriendName.c_str());
+      const auto nFileNames = friendFileNames[i].size();
+      if (thisFriendChainSubNames.empty()) {
+         // If there are no chain subnames, the friend was a TTree. It's safe
+         // to add to the chain the filename directly.
+         for (auto j = 0u; j < nFileNames; ++j) {
+            frChain->Add(thisFriendFiles[j].c_str());
+         }
+      } else {
+         // Otherwise, the new friend chain needs to be built using the nomenclature
+         // "filename/treename" as argument to `TChain::Add`
+         for (auto j = 0u; j < nFileNames; ++j) {
+            frChain->Add((thisFriendFiles[j] + "?#" + thisFriendChainSubNames[j]).c_str());
+         }
+      }
+
+      // Make it friends with the main chain
+      fTree->AddFriend(frChain.get(), thisFriendAlias.c_str());
+      // fFriends are needed to preserve the friends alive after the end of the ctor call
+      fFriends.emplace_back(std::move(frChain));
+   }
```

Comment:
In fact, this is the same code as in https://github.com/root-project/root/blob/dea97f02bdfaf788f2c96ae2e02d20a64744364d/tree/treeplayer/src/TTreeProcessorMT.cxx#L318-L345.

Afaik, @vepadulano is working on understanding the lifetime of the friends, so that we move the quoted logic to a separate function.

Reply:
Yes, I am convinced that the signature written by Enrico above is the only feasible one. There are unfortunately quite huge differences between C++ and Python lifetime management, but the signature above should work with both in any case.